src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -124,50 +173,75 @@ absl::optional<int64_t> GcsTaskManager::GcsTaskManagerStorage::GetTaskStatusUpda
              : absl::nullopt;
 }
 
+void GcsTaskManager::GcsTaskManagerStorage::MarkTasksFailed(const JobID &job_id,
+                                                            int64_t job_finish_time_ns) {
+  auto task_attempts_itr = job_to_task_attempt_index_.find(job_id);
+  if (task_attempts_itr == job_to_task_attempt_index_.end()) {
+    // No tasks in the job.
+    return;
+  }
+
+  // Iterate all task attempts from the job.
+  for (const auto &task_attempt : task_attempts_itr->second) {
+    if (!IsTaskTerminated(task_attempt.first)) {
+      MarkTaskAttemptFailed(task_attempt, job_finish_time_ns);
+    }
+  }
+}
+
 void GcsTaskManager::GcsTaskManagerStorage::MarkTaskFailed(const TaskID &task_id,
                                                            int64_t failed_ts) {
   auto latest_task_attempt = GetLatestTaskAttempt(task_id);
   if (!latest_task_attempt.has_value()) {
     return;
   }
-  auto &task_event = GetTaskEvent(*latest_task_attempt);
-  task_event.mutable_state_updates()->set_failed_ts(failed_ts);
-  task_event.mutable_state_updates()->clear_finished_ts();
+  MarkTaskAttemptFailed(*latest_task_attempt, failed_ts);
 }
 
 void GcsTaskManager::GcsTaskManagerStorage::MarkTaskTreeFailedIfNeeded(
     const TaskID &task_id, const TaskID &parent_task_id) {
+  absl::optional<int64_t> ancestor_failed_ts = absl::nullopt;
```

Comment:
I found this logic a bit hard to follow. Do you see any con if we use recursion here? E.g.,

```
def FailTaskTree(task_id):
    children = get_children(task_id)
    task = get_task_attempt(task_id)

    if (task.state == FAILED):
        # meaning all the children has failed
        return

    task.state = Failed
    for child in children:
        FailTaskTree(child.task_id)
```
?

Reply:
> As for the pseudocode you have above, however, would still override a finished child task's state to fail?

Ah I think the exit state of recursion could be 

```
if (task.state == FAILED || task.state == FINISHED):
```
> Also, the above routine will still need to check the parent's failure. So the root is actually different from the children, the root being a failure implies we need to mark potential children as failure, while the children being failure means we could stop at the child.

Hmm is it true though? I assume

1. If the parent has already failed, this child should've been marked as failed. So it will reach to `if (task.state == FAILED || task.state == FINISHED):` and stop recursion. 
2. If the parent hasn't failed yet, it will mark all children as failed. Whe the parent failed, it will check the current node and see "state == FINISHED or FAILED", meaning recursion will stop there

> If it's hard to follow, I should probably do a better job documenting it. What are the parts that are not clear?

I just feel like there's a lot of logic compared to the recursion solution. Mainly I don't 100% understand why we need to check the ancestor timestamp (as I mentioned above ^), but I might be missing something.

Reply:
Even with `task.state ==FINISHED` check,  the recursion will stop incorrectly if there is an intermediate task that finishes. 
E.g. 
`A -> B -> C`
1. All of them running 
2. B finishes (while it is not the most normal use cases but could happen) 
3. A fails

We should mark C as failed in this case? But w/o additional info like ancestor failed ts, there's no way just from failed/finished timestamp to reach C.

Reply:
Hmm interestingly when I tried your scenario in real ray, C is still running (even after A fails). 

```python3
In [3]: @ray.remote
   ...: def f():
   ...:     pass
   ...: 

In [4]: @ray.remote
   ...: def f():
   ...:     import time
   ...:     time.sleep(300)
   ...: 

In [5]: @ray.remote
   ...: def child():
   ...:     a = f.remote()
   ...: 

In [6]: @ray.remote
   ...: def parent():
   ...:     ray.get(child.remote())
   ...:     import time
   ...:     time.sleep(300)
   ...: 

In [7]: p = parent.remote()
```
this is what I tried (and I manually killed parent ps)

And more interestingly, even after I killed the job, f still keeps running...

Reply:
(lol this is why we thought this approach is a bit sketchy... we don't know the exact high level behavior) 

Maybe for these, we should write an integration test to verify if the hypothesis & real scenario is the same. It looks like the grandparent failure doesn't affect the grandchild's failure if the parent didn't fail

Reply:
(which also means we probably don't need to handle this ancestor failure? )

Reply:
Ah - great find! I took it for granted on the grandparent - parent - child failure model...

Seems the child will become an orphan if the parent has already finished.  Not sure if it's the most intuitive semantics but yeah, I agree we should probably not handling ancestor failure.